day8/alienalphabet.py

```diff
@@ -0,0 +1,131 @@
+##############
+## Solution ##
+##############
+
+def alienAlphabet(orderedWords):
+    # edge cases
+    if len(orderedWords) <= 0:
+        return ''
+    if len(orderedWords) == 1:
+        return removeDups(orderedWords[0])
+
+    # linearization of a DAG
+    graph = createDependencyGraph(orderedWords)
+    alphabet = topSort(graph)
+
+    return "".join(alphabet)
+
+#############
+## Helpers ##
+#############
+
+def createDependencyGraph(orderedWords):
+    graph = {}
+    for i, word in enumerate(orderedWords[:-1]):
+        nextWord = orderedWords[i+1]
+        shorter = min(len(word), len(nextWord))
+        for j in xrange(shorter):
+            letter = word[j]
+            otherLetter = nextWord[j]
+
+            if letter not in graph:
+                graph[letter] = []
+            if otherLetter not in graph:
+                graph[otherLetter] = []
+
+            if letter != otherLetter:
+                graph[letter].append(otherLetter)
+                break
+
+    return graph
+
+# graph is represented as key:value pair
+# key = node
+# value = [list, of, connected, nodes]
+#
+# the graph is also given to be a DAG so all the connected
+# nodes are from outgoing edges
+#
+# this function returns the resulting linearization in a list
+def topSort(graph):
+    sources = getSources(graph)
+    visited = set()
+    postOrderStack = []
+
+    def dfs(start, graph, visitedSet, postOrderStack):
+        visitedSet.add(start)
+
+        # recurse
+        for child in graph[start]:
+            if child not in visitedSet:
+                dfs(child, graph, visitedSet, postOrderStack)
+
+        # done exploring your children, add yourself to the postorder
+        postOrderStack.append(start)        
+
+    # run dfs from all source nodes
+    for node in sources:
+        dfs(node, graph, visited, postOrderStack)
+
+    return postOrderStack[::-1]
+
+def getSources(graph):
+    # initialize all nodes to 0
+    nodes = {k:0 for k in graph.keys()}
```

Comment:
you're using `nodes` as a set (using values of 0 or 1 for measuring belonging). you could have just done something like

    non_source_nodes = set()
    for target_node_list in graph.values():
        for node in target_node_list:
            non_source_nodes.add(node)
    source_nodes = set(graph.keys()) - non_source_nodes
    return source_nodes

imo this makes it more clear what the dictionary/set represents, because your 1s and 0s are only explained in comments